cryptography-providers/openssl3/api/src/commonMain/kotlin/algorithms/Openssl3AesCmac.kt

```diff
@@ -0,0 +1,153 @@
+private class AesCmacSignature(
+    private val algorithm: String,
+    private val key: ByteArray,
+) : SignatureGenerator, SignatureVerifier {
+    private val mac = checkError(EVP_MAC_fetch(null, "CMAC", null))
+    private fun createFunction() = AesCmacFunction(
+        key = key,
+        algorithm = algorithm,
+        context = Resource(checkError(EVP_MAC_CTX_new(mac)), ::EVP_MAC_CTX_free)
+    )
+
+    override fun createSignFunction(): SignFunction = createFunction()
+    override fun createVerifyFunction(): VerifyFunction = createFunction()
+
+    private class AesCmacFunction(
+        private val key: ByteArray,
+        private val algorithm: String,
+        private val context: Resource<CPointer<EVP_MAC_CTX>>,
+    ) : SignFunction, VerifyFunction, SafeCloseable(SafeCloseAction(context, AutoCloseable::close)) {
+
+        init {
+            reset()
+        }
+
+        @OptIn(UnsafeNumber::class)
+        override fun update(source: ByteArray, startIndex: Int, endIndex: Int) {
+            // Implementation for updating the CMAC with the provided data
+            checkBounds(source.size, startIndex, endIndex)
+            val context = context.access()
+            source.usePinned {
+                checkError(
+                    EVP_MAC_update(
+                        ctx = context,
+                        data = it.safeAddressOf(0).reinterpret(),
+                        datalen = source.size.convert()
+                    )
+                )
+            }
+        }
+
+        override fun signIntoByteArray(destination: ByteArray, destinationOffset: Int): Int {
+            val signature = signToByteArray()
+            checkBounds(destination.size, destinationOffset, destinationOffset + signature.size)
+            signature.copyInto(destination, destinationOffset, destinationOffset)
+            return signature.size
+        }
+
+        @OptIn(UnsafeNumber::class)
+        override fun signToByteArray(): ByteArray {
+            val context = context.access()
+            val macSize = EVP_MAC_CTX_get_mac_size(context).convert<Int>()
+            val out = ByteArray(macSize)
+            out.usePinned {
+                checkError(
+                    EVP_MAC_final(
+                        ctx = context,
+                        out = it.safeAddressOf(0).reinterpret(),
+                        outl = null,
+                        outsize = macSize.convert()
+                    )
+                )
+            }
+            return out
+        }
+
+        @OptIn(UnsafeNumber::class)
+        override fun reset() {
+            val context = context.access()
+            memScoped {
+                val params = allocArrayOf(
+                    OSSL_PARAM_construct_utf8_string(
+                        "cipher".cstr.ptr,
+                        algorithm.cstr.ptr,
+                        0.convert()
+                    ),
+                    OSSL_PARAM_construct_end()
+                )
```

Comment:
It's possible to simplify it a bit using helper function:
```suggestion
                val params = OSSL_PARAM_array(
                    OSSL_PARAM_construct_utf8_string(
                        "cipher".cstr.ptr,
                        algorithm.cstr.ptr,
                        0.convert()
                    )
                )
```